File: 0086-partition-list/0086-partition-list.js
var partition = function(head, x) {
    // Create two dummy nodes for the two partitions.
    let lessHead = new ListNode(0);
    let greaterHead = new ListNode(0);
    
    // Create two pointers to iterate over the two lists.
    let less = lessHead;
    let greater = greaterHead;
    
    // Traverse the original list.
    while (head !== null) {
        if (head.val < x) {
            less.next = head;  // Add the node to the 'less' list.
            less = less.next;
        } else {
            greater.next = head;  // Add the node to the 'greater' list.
            greater = greater.next;
        }
        head = head.next;
    }
    
    // End the 'greater' list to prevent cycle.
    greater.next = null;
    
    // Link the two lists together.
    less.next = greaterHead.next;
    
    // Return the head of the 'less' list.
    return lessHead.next;
};

File: 0086-partition-list/0086-partition-list.cpp
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        // Dummy nodes to start the two lists
        ListNode* before_head = new ListNode(0); // Nodes less than x
        ListNode* after_head = new ListNode(0);  // Nodes greater than or equal to x

        // Pointers to the current node in each list
        ListNode* before = before_head;
        ListNode* after = after_head;

        // Traverse the original list
        while (head != nullptr) {
            if (head->val < x) {
                // Add to the 'before' list
                before->next = head;
                before = before->next;
            } else {
                // Add to the 'after' list
                after->next = head;
                after = after->next;
            }
            head = head->next;
        }

        // Link the two lists together
        after->next = nullptr;  // End the after list
        before->next = after_head->next;  // Append after list to before list

        // Return the head of the modified list
        return before_head->next;
    }
};
